### PR TITLE
Dev->main

### PR DESCRIPTION
**Pull request by Cartmanezs**

### CODE REVIEW

File: FruitsAndBerries/Main/FruitsAndBerriesInteractor.swift [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    
+    init(fruitsBerriesRepository: FruitsFetchRepository) {
+        self.fruitsBerriesRepository = fruitsBerriesRepository
+    }
+
+    private func fetchData() {
+        fruitsBerriesRepository.fetchRandom { result in
+            switch result {
+            case .success(let itemsResponse):
+                let dispatchGroup = DispatchGroup()
```
Comment (valself) [inline FruitsAndBerries/Main/FruitsAndBerriesInteractor.swift:29]: Looks too complex!
Why do you use dispatchgroups here? You call method that already has a completion. Unwrap this completion and handle it. That's all. 

Read about KISS principle.

File: FruitsAndBerries/Main/FruitsAndBerriesInteractor.swift [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                
+                let responseItems = itemsResponse.items.compactMap { item in
+                    let imageUrlString = item?.image == nil ? "" : "\((item!.image)!)"
+                    dispatchGroup.enter()
+                    var image: UIImage?
+                    self.fruitsBerriesRepository.fetchImage(imageUrlString) { resultImage in
```
Comment (valself) [inline FruitsAndBerries/Main/FruitsAndBerriesInteractor.swift:35]: memory leak!

File: FruitsAndBerries/Main/FruitsAndBerriesInteractor.swift [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            }
+        }
+    }
+    
+    private func fetchDetails(for itemId: String) {        
+        fruitsBerriesRepository.fetchDetailsInfo(itemId: itemId) { result in
+            switch result {
+            case .success(let detailsResponse):
+                let loadDetailsResponse = FruitsAndBerriesItemsDataModel.LoadDetails.Response(itemDetail: detailsResponse)
+                self.presenter?.presentDetails(response: loadDetailsResponse)
```
Comment (valself) [inline FruitsAndBerries/Main/FruitsAndBerriesInteractor.swift:63]: leak

File: FruitsAndBerries/Main/FruitsAndBerriesInteractor.swift [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            case .failure(let error):
+                print("Error fetching details: \(error)")
```
Comment (valself) [inline FruitsAndBerries/Main/FruitsAndBerriesInteractor.swift:65]: how user will be notified about this case?

File: FruitsAndBerries/Main/FruitsAndBerriesInteractor.swift [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        self.fruitsBerriesRepository = fruitsBerriesRepository
+    }
+
+    private func fetchData() {
+        fruitsBerriesRepository.fetchRandom { result in
+            switch result {
+            case .success(let itemsResponse):
+                let dispatchGroup = DispatchGroup()
+                
+                let responseItems = itemsResponse.items.compactMap { item in
```
Comment (valself) [inline FruitsAndBerries/Main/FruitsAndBerriesInteractor.swift:31]: Looks too complex again... Find more simple solution

**Review by valself — COMMENTED** [review]
I'm waiting for the changes

File: FruitsAndBerries/Detail/FruitsAndBerriesDetailsViewController.swift [hunk 6]
```diff
@@ -0,0 +1,129 @@
+//
+//  FruitsAndBerriesDetailsViewController.swift
+//  FruitsAndBerries
+//
+//  Created by Igor Palyvoda on 03.01.2024.
+//
+
+import UIKit
+
+final class FruitsAndBerriesDetailsViewController: UIViewController {
```
[the remark this answers is not in the document]
Comment (valself) [reply FruitsAndBerries/Detail/FruitsAndBerriesDetailsViewController.swift:10]: Ok I see this text in the requirements

`Although 2 screens need to be implemented, they should both be a part of the same module. Navigating between them should happen completely in the View layer (not through the router) by simply pushing the details screen (View Controller to be created) into the navigation controller.`

File: FruitsAndBerries/API/Endpoints.swift [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        )
+    }
+
+    static func detailsInfo(itemId: String) -> Self {
+        Endpoint(
+            path: "/texts/\(itemId)"
+        )
+    }
+    
+    static func image(imageUrl: String) -> Self {
```
[the remark this answers is not in the document]
Comment (Cartmanezs) [reply FruitsAndBerries/API/Endpoints.swift:29]: Changed to imageUrlString

File: FruitsAndBerries/API/DError.swift [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+//  FruitsAndBerries
+//
+//  Created by Igor Palyvoda on 04.01.2024.
+//
+
+import Foundation
+
+// MARK: Initialization
+
+enum DError: Error {
```
[the remark this answers is not in the document]
Comment (Cartmanezs) [reply FruitsAndBerries/API/DError.swift:12]: Changed to FBError -  FruitsAndBerrieError

[the remark this answers is not in the document]
Comment (valself) [reply FruitsAndBerries/API/DError.swift]: You didn't change this name in the header

File: FruitsAndBerries/API/Endpoints.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (valself) [reply FruitsAndBerries/API/Endpoints.swift]: > Changed to imageUrlString

I don't see it

[the remark this answers is not in the document]
Comment (Cartmanezs) [reply FruitsAndBerries/API/Endpoints.swift]: for now it's imageUrl: URL

File: FruitsAndBerries/Repositories/FruitsAndBerrieRepository.swift [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+         self.httpClient = URLSessionHTTPClient(session: session)
+     }
+}
+
+// MARK: Private Methods
+
+private extension FruitsAndBerrieRepository {
+    func fetch<T>(with endpoint: Endpoint, completion: @escaping (Result<T, DError>) -> Void) where T: Decodable {
+        guard let url = endpoint.url else {
+            preconditionFailure("Unable to construct a valid URL for given endpoint: \(endpoint)")
```
[the remark this answers is not in the document]
Comment (Cartmanezs) [reply FruitsAndBerries/Repositories/FruitsAndBerrieRepository.swift]: I've already checked. It doesn't crash.

**Review by valself — CHANGES_REQUESTED** [review]

**Pull request closed without merging**